src/core/pratt.ml

```diff
@@ -0,0 +1,113 @@
+(** This module defines functions needed by the Pratt parser of the Pratter
+    library. *)
+
+open Syntax
+
+module Pratt : sig
+  val parse : Sig_state.t -> Env.t -> p_term -> p_term
+end = struct
+
+  open Lplib
+  open Extra
+  open Pos
+
+  (** [find_qid ss env qid] fetches the qualified identifier associated to
+      [qid] if [qid] is the identifier of an infix operator defined in
+      signature state [ss]. If [qid] is defined in environment [env], then it
+      designates a variable, and is thus returned as-is. *)
+  (* REVIEW: this function relies on a lot of invariants and performs many
+     [Map.find_opt]s. Things may be optimised by modifying the data structure
+     of the signature state, notably making the qualified identfier to which
+     operators are linked available in the [unops] field of the signature
+     state. *)
+  let find_qid : Sig_state.t -> Env.t -> qident -> qident =
+    fun ss env ({elt=(mp, s); _} as qid) ->
+    try
+      if mp <> [] then raise Not_found;
+      ignore (Env.find s env);
+      qid
+    with Not_found ->
+    match StrMap.find_opt s ss.Sig_state.unops with
+    | Some sym -> (
+        match Terms.SymMap.find_opt sym ss.Sig_state.pp_hints with
+        | Some (Prefix(_, _, qid)) -> qid
+        | _ -> assert false )
+    | None -> (
+        match StrMap.find_opt s ss.Sig_state.binops with
+        | Some sym -> (
+            match Terms.SymMap.find_opt sym ss.Sig_state.pp_hints with
+            | Some (Infix (_, _, _, qid)) -> qid
+            | _ -> assert false )
+        | None -> qid )
+
+
+  module Pratt_terms : Pratter.SUPPORT
+    with type term = p_term
+     and type table = Sig_state.t * Env.t
+  = struct
+    type term = p_term
+    type table = Sig_state.t * Env.t
+
+    let get_binary (tbl, _) t =
+      let open Sig_state in
+      match t.elt with
+      | P_Iden(id, _) -> (
+          let {elt=(_, suffix); _} = id in
+          match StrMap.find_opt suffix tbl.binops with
+          | None -> None
+          | Some(sym) ->
+          match Terms.SymMap.find_opt sym tbl.pp_hints with
+          | Some(Infix((_, assoc, prio, _))) -> (
+              let assoc =
+                match assoc with
+                | Assoc_left -> Pratter.Left
+                | Assoc_right -> Pratter.Right
+                | Assoc_none -> assert false
```

Comment:
Non-associative infix symbols like `::` (cons) on lists are important and useful. We need to handle them. I don't understand why this is a problem. Could you please elaborate on that?

Reply:
Well, `::` is perhaps not a very good example as it could be considered as right associative: 1::2::[] = 1::(2::[]).
If you have c:T->U->V and want x c y c z = x c (y c z), you need x:T, y:T, z:U and V=U, so c:T->U->U.
Conversely, if you want x c y c z = (x c y) c z, you need x:T, y:U, z:U and V=T, so c:T->U->T.

Reply:
Files where you mentioned a problem:
rewrite.lp: = is declared as infix; no associativity is possible here; yet this notation is VERY important! ;-)
328.lp: `::` declared infix with no associativity (could be fixed by declaring it right associative)
builtin_zero_succ.lp: + is declared infix with no associativity (yet); note that someone may want to declare an infix notation, later prove that it is associative and add associativity as a rule?
inductive_type.lp: = is declared infix like in rewrite.lp
(I cannot find the files printA.lp, printB.lp and printC.lp that you also mentioned.)

Reply:
> Well, `::` is perhaps not a very good example as it could be considered as right associative: 1::2::[] = 1::(2::[]).
> If you have c:T->U->V and want x c y c z = x c (y c z), you need x:T, y:T, z:U and V=U, so c:T->U->U.
> Conversely, if you want x c y c z = (x c y) c z, you need x:T, y:U, z:U and V=T, so c:T->U->T.

But what should be the behaviour of non associative symbols? In the previous parser, parsing failed on an input `x c y c z`. Do we keep this? When you say `x c y c z = x c (y c z)`, `c` is right associative, right?

Reply:
Yes. If c is not associative, x c y c z should generate a syntax error.